File: Leetcode/PY/Leetcode846-一手顺子.py
class Solution:
    def isNStraightHand(self, hand: List[int], groupSize: int) -> bool:
        st,n = defaultdict(int),len(hand)
        if n % groupSize != 0: return False
        hand.sort()
        for hd in hand:
            st[hd] += 1
        for hd in hand:
            if st[hd] == 0: continue
            for t in range(hd,hd + groupSize):
                if st[t] == 0: return False
                st[t] -= 1
        return True
         
        

File: Leetcode/JS/Leetcode846-一手顺子.js
var isNStraightHand = function(hand, groupSize) {
    const n = hand.length;
    if(n % groupSize != 0) return false;
    const st = new Map();
    hand.sort((a,b) => a - b);
    for(const hd of hand){
        st.set(hd,(st.get(hd) || 0) + 1);
    }
    for(const hd of hand){
        if(!st.has(hd)) continue;
        for(let i = hd;i < hd + groupSize;i++){
            if(!st.has(i)) return false;
            st.set(i,st.get(i) - 1);
            if(st.get(i) == 0) st.delete(i);
        }
    }
    return true;
};